### PR TITLE
[#99121602] broker rennovate

### PR DESCRIPTION
**Pull request by dhilton**
__What__

A renovation of the tsuru Elasticsearch broker to add test coverage, basic auth support and to enable it to run as a tsuru app, with updated configuration for the latest service broker specification.

This was done as part of https://www.pivotaltracker.com/story/show/99121602 

__How to Review this PR__

This PR should be reviewed in the following context:

**Given** that there is a platform running
**When** the performance scripts are run
**Then** all the deployed applications point to a ElasticSearch server(s) hosted on the platform
**And** the ElasticSearch databases are all populated with realistic data

The scope is purely to support the performance tests, not to solve all possible needs or problems with the moving this into a production quality state. 

To test this PR, create a new environment, folllow the documentation for creating a new service using crane in that environment and try binding a service instance to an app instance. We used the digital marketplace search api app for this, however to fully test the ES integration you'll need an elasticsearch server. However, to test this PR you just need to check that the correct environment variables have been set in your app-instance's containers. That proves that the service works correctly with tsuru-api and injects the correct environment variables.

__Who should review this PR__

Any one in the team except @mtekel and @RichardKnop  who both worked on this codebase as well as I.

### CODE REVIEW

File: api.py [hunk 1]
```diff
@@ -1,28 +1,88 @@
...
+def remove_instance(name):
+    return "", 200
+
+
+@app.route("/resources/<name>/bind-app", methods=["POST"])
+def bind_app(name):
+    if ELASTICSEARCH_IP:
+        ELASTICSEARCH_HOST = ELASTICSEARCH_IP
+    else:
+        ELASTICSEARCH_HOST = get_ip_address('eth0')
```
Comment (dcarley) [inline api.py]: I don't understand what this is used for and it's not described by the commit, README, or tests. Does it return the IP of the container that the broker is running on? Why would an Elasticsearch server be running there?

File: api.py [hunk 2]
```diff
@@ -1,28 +1,88 @@
...
+ELASTICSEARCH_PORT = os.environ.get("ELASTICSEARCH_PORT", '9200')
 
-@bottle.route("/resources/<name>", method="POST")
+def get_ip_address(ifname):
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    return socket.inet_ntoa(fcntl.ioctl(
+        s.fileno(),
+        0x8915,  # SIOCGIFADDR
+        struct.pack('256s', ifname[:15])
+    )[20:24])
```
Comment (dcarley) [inline api.py]: What are the magic numbers on these three lines? Can you define descriptively named constants for them instead?

Comment (dhilton) [inline api.py]: Yeah - this is a bit voodoo like, will simplify. It's a python trick to get the IP.

http://code.activestate.com/recipes/439094-get-the-ip-address-associated-with-a-network-inter/

File: api.py [hunk 3]
```diff
@@ -1,28 +1,88 @@
...
+@app.route("/resources/<name>/bind", methods=["POST"])
 def bind(name):
-    bottle.response.status = 201
-    return json.dumps({"ELASTIC_SEARCH_URL": ELASTIC_SEARCH_URL})
+    return "", 201
+
+
+@app.route("/resources/<name>/status", methods=["GET"])
+def status(name):
+    health_check_url = "http://{0}:{1}/_cluster/health?pretty=true".format(ELASTICSEARCH_IP, ELASTICSEARCH_PORT)
```
Comment (dcarley) [inline api.py]: Related to https://github.com/alphagov/elastic-search-api/pull/1/files#r35302163 - is it valid to use the broker without `ELASTICSEARCH_IP` set? In which case this won't form a valid URL.

Accordingly this new endpoint probably deserves some basic tests.

File: api.py (hunk 2 above)
Comment (jimconner) [inline api.py]: The 0x8915 value is the IOCTL call to the linux netdevice to trigger the get-address function

 SIOCGIFADDR, SIOCSIFADDR
   Get  or set the address of the device using ifr_addr.  Setting the  interface  address  is  a  privileged   operation.  For compatibility,   only   AF_INET   addresses  are  accepted  or  returned.

File: api.py [hunk 4]
```diff
@@ -1,28 +1,88 @@
...
+    es_state = requests.get(health_check_url)
+    if es_state.status_code == 200:
```
Comment (dcarley) [inline api.py]: Is this status code an indication of whether Elasticsearch is healthy or more simply whether we could get a response from Elasticsearch? I thought the former at first, but I'm not sure that the Elasticsearch endpoint returns a non-200 response code when the cluster is non-green. A couple of tests with descriptive names might make it clearer.

In the case of a non-200 response, should we also provide the body that Elasticsearch gave us to help diagnose any problems? Should we pass on the response verbatim?

**Comment by dcarley** [comment]
:+1:

**Pull request merged**